### ISSUE TITLE
[API Proposal]: `[LibraryExport]` analogous to `[JSExport]`

### ISSUE THREAD

**Issue by am11** [dotnet/runtime#79786]
### Background and motivation

In .NET 7 we have:

* `[JSImport]` and `[JSExport]` to import and export JavaScript entrypoints
* `[LibraryImport]` and `[UnmanagedCallersOnly]` to import and export native entrypoints with specified calling convention

From this outlook, `[UnmanagedCallersOnly]` name is unsymmetrical and the odd one out. 

`LibraryExport` as attribute name would have been much cleaner, especially for AOT scenarios. Aside from the naming:

* "exporting native entrypiont" is not the only use-case of `UnmanagedCallersOnlyAttribute`; it was originally added for  roslyn to support `unmanaged` semantics.
* it has a non-intuitive restriction for "exporting native entrypoint" that it doesn't work  without the explicit `EntryPoint` value.

There is a room to add a dedicated `LibraryExportAttribute` with a single intent, similar to `JSExportAttribute`.

This may also give a chance for runtimes to perform targeted optimizations (reduce # of frames around these calls in coreclr) with a slight ease than it is today.

### API Proposal

```c#
namespace System.Runtime.InteropServices
{
    [AttributeUsage(AttributeTargets.Method, Inherited = false)]
    public class LibraryExportAttribute : Attribute
    {
        public LibraryExportAttribute() { }
        public Type[]? CallConvs; // optional; defaults to current platform's calling convention
        public string? EntryPoint; // unlike UnmanagedCallersOnly, this is optional for native export
    }
}
```

### API Usage

```c#
    [LibraryExport]
    public static int GetDotnetMajorVersion() => Environment.Version.Major;
```

test.c
```c
#include<stdio.h>

extern int GetDotnetMajorVersion(void);

int main()
{
  printf(".NET major version is: %d\n", GetDotnetMajorVersion());
  return GetDotnetMajorVersion();
}
```

```sh
$ dotnet publish -r linux-x64 -p:PublishAot=true -p:NativeLib=Shared
$ cc test.c bin/Debug/net7.0/linux-x64/publish/mynativelib1.so

$ ./a.out
.NET major version is: 7

$ echo $?
7
```

### Risks

Confuse the consumers who have made their peace with `UnmanagedCallersOnly(EntryPoint="mandatory_name")`.

**Comment by teo-tsirpanis** [comment]
For symmetry with `LibraryImport`, `LibraryExport` should not have a `CallConvs` property and instead rely on the `UnmanagedCallConvAttribute` if exists.

Another thing to consider is to incorporate marshalling in the design of `LibraryExport`, and support exposing the generated blittable `UnmanagedCallersOnly` function to the user, so that we would be able to write something like this:

```csharp
public unsafe partial class MyClass
{
    // GenerateNativeThunk to true would cause the source generator to emit a Callback_native
    // method with UnmanagedCallersOnly inside the declaring class, instead of a file class
    // to be seen by just NativeAOT.
    [LibraryExport(StringMarshalling = StringMarshalling.Utf16, GenerateNativeThunk = true)]
    private static void Callback(string s)
    {
        Console.WriteLine(s);
    }

    [LibraryImport("mylibrary")]
    private static void MyPInvoke(delegate* unmanaged<char*, void> callback);

    static void Main()
    {
        MyPInvoke(&Callback_native);
    }
}
```

**Comment by am11** [comment]
> `LibraryExport` should not have a `CallConvs` property and instead rely on the `UnmanagedCallConvAttribute` if exists.

Sounds reasonable. I haven't put much thought into `CallConvs` and source generator, but if we can seamlessly resolve the user intent in source generator without `CallConvs` property, that would be a preferred shape for this attribute.

Tough I'm not sure if lowering `LibraryExportAttribute` to `UnmanagedCallConvAttribute`  in source generator is what we would ultimately want in a long run, as these models are relatively new (e.g. its support in Mono AOT is under development https://github.com/dotnet/runtime/issues/79377).

**Comment by teo-tsirpanis** [comment]
As I'm imagining it, `[LibraryExport]` would be the high-level equivalent of `[UnmanagedCallersOnly]` just like `[LibraryImport]` would be the high-level equivalent of `[DllImport]`. The former would support source-generated marshalling and the latter wouldn't.

**Comment by AaronRobinsonMSFT** [comment]
I appreciate the symmetrically argument but I don't think this is appropriate. This is unfortunately an area where there isn't a clean separation and we would like to not introduce yet more confusion. The history here is tangled up in NativeAOT as well where we had an attribute that was similar - that is why we added the `EntryPoint` to `UnmanagedCallersOnly`. The `UnmanagedCallersOnly` already handles all needed scenarios, both internal and exporting and I don't see a lot of value in adding another just for symmetry sake. I will keep this on 8 for consideration but will likely close this toward the end of the release without a reason more compelling than "symmetry".

**Comment by jkoritzinsky** [comment]
I think this would be a good feature to add and I would put this in a similar category as #63590.

Today, we don't have a good mechanism for both providing unmanaged-callable functions and allowing marshalling without using the built-in marshalling system. This feature could be used as a counterpart to #63590 to enable implementing a fully-source-generated interop story for "delegate marshalling"-style scenarios.

Additionally, I think this feature actually fits well with our COM source-generation direction work, although possibly not the work that will make it into .NET 8 in particular. With the COM generator, we'll have support for both managed->unmanaged and unmanaged->managed stubs for implementing source-generated CCW/RCW implementations that support all of the goodness of the new source-generation marshalling model we produced in .NET 7. However, to implement a COM server (a la comhost but for source-generated COM interop) entirely with source-generated code, a developer would still have to manually write out the marshalling for any parameters in the API the COM server must implement. This would be a weird corner where we'd have a hole in our abstraction model.

**Comment by AaronRobinsonMSFT** [comment]
> `GenerateNativeThunk = true`

@teo-tsirpanis I don't get this scenario. For the "export" scenario, under what circumstances would we not generate the `UnmanagedCallersOnly`?

**Comment by jkoritzinsky** [comment]
> > `GenerateNativeThunk = true`
> 
> @teo-tsirpanis I don't get this scenario. For the "export" scenario, under what circumstances would we not generate the `UnmanagedCallersOnly`?

The idea is when `GenerateNativeThunk=false`, the generated method is in a `file`-scoped class and is only exposed via a mechanism that supports projecting an `UnmanagedCallersOnlyAttribute` with the `EntryPoint` property set as a native export (i.e. DNNE, NativeAOT, MonoAOT starting in .NET 8). In this case, the user wouldn't be able to access the stub directly, but it would be exposed to native callers.

When `GenerateNativeThunk=true`, the generated method would be in a type that is accessible from user-written code, so the user could write `&ManagedMethod_UnmanagedWrapper` in their C# and get the unmanaged function pointer to the method.

I don't know if this scenario is compelling enough to provide the two different options, and I think the `GenerateNativeThunk=true` scenario is more compelling at first glance.

**Comment by am11** [comment]
`UnmanagedCallersOnly` to me looks like a multi-purpose attribute; started off as a compiler support attribute and ended up supporting other scenarios like NativeAOT exports. I was particularly looking at the poor codegen around the exported entrypoints (which are probably hard to optimize with the current `UnmanagedCallersOnly` machinery). With `LibraryExport` new surface area, I was hoping if it could be designed in a way that the final codegen would shave off _some_ of the instructions in blittable / primitive scenarios (pertaining to GC, P/Invoke transitions etc.).

> I don't know if this scenario is compelling enough to provide the two different options, and I think the `GenerateNativeThunk=true` scenario is more compelling at first glance.

Provided generator will emit linker-friendly code, linker will drop the unused code from wrapping type for export-only usages. With that in mind, I agree that an additional switch might not be necessary.

**Comment by jkotas** [comment]
Where do you see inefficiencies in the existing `UnmanagedCallersOnly` that can be optimized? `UnmanagedCallersOnly` optimization should be independent on this proposal.

**Comment by am11** [comment]
@jkotas, agreed. This discussion has evolved to source generator which will emit `UnmanagedCallersOnly` under the hood and no longer a independent / parallel API to `UnmanagedCallersOnly`.

> Where do you see inefficiencies in the existing `UnmanagedCallersOnly` that can be optimized?

I was initially looking at simple exports (involving primitive constructs and constants), e.g. with:
```c#
using System.Runtime.InteropServices;
public class C
{
    [UnmanagedCallersOnly(EntryPoint = nameof(M))]
    public static int M() => 42;
}
```
on linux-arm64, we get:
```asm
// function con42_C__M

stp	x29, x30, [sp, #-48]!
str	x19, [sp, #40]
mov	x29, sp
add	x0, x29, #0x18
bl	0x15d40 <RhpReversePInvoke(ReversePInvokeFrame*)>
mov	w19, #0x2a                  	// #42
add	x0, x29, #0x18
bl	0x15e14 <RhpReversePInvokeReturn(ReversePInvokeFrame*)>
mov	w0, w19
ldr	x19, [sp, #40]
ldp	x29, x30, [sp], #48
ret
```

I was thinking perhaps for simple cases like this one, it is possible to skip `ReversePInvokeFrame` and optimize it to inlineable `return 42`.

**Comment by am11** [comment]
@jkotas, agreed. This discussion has evolved to source generator which will emit `UnmanagedCallersOnly` under the hood and no longer an independent / parallel API to `UnmanagedCallersOnly`.

> Where do you see inefficiencies in the existing `UnmanagedCallersOnly` that can be optimized?

I was initially looking at simple exports (involving primitive constructs and constants), e.g. with:
```c#
using System.Runtime.InteropServices;
public class C
{
    [UnmanagedCallersOnly(EntryPoint = nameof(M))]
    public static int M() => 42;
}
```
on linux-arm64, we get:
```asm
// function con42_C__M

stp	x29, x30, [sp, #-48]!
str	x19, [sp, #40]
mov	x29, sp
add	x0, x29, #0x18
bl	0x15d40 <RhpReversePInvoke(ReversePInvokeFrame*)>
mov	w19, #0x2a                  	// #42
add	x0, x29, #0x18
bl	0x15e14 <RhpReversePInvokeReturn(ReversePInvokeFrame*)>
mov	w0, w19
ldr	x19, [sp, #40]
ldp	x29, x30, [sp], #48
ret
```

I was thinking perhaps for simple cases like this one, it is possible to skip `ReversePInvokeFrame` and optimize it to inlineable `return 42`.

**Comment by AaronRobinsonMSFT** [comment]
> When GenerateNativeThunk=true, the generated method would be in a type that is accessible from user-written code, so the user could write &ManagedMethod_UnmanagedWrapper in their C# and get the unmanaged function pointer to the method.

I see. That seems like we are getting into more and more bespoke scenarios. However, the mechanism I believe would be simply to not set the `EntryPoint` field, right? If so, the cost would be quite small to provide the mechanism and avoid creating an "export". I could be mistaken but I though creating an "export" required setting the `EntryPoint`, so creating a file-scoped declaration wouldn't be needed.

> I don't know if this scenario is compelling enough to provide the two different options, and I think the GenerateNativeThunk=true scenario is more compelling at first glance.

Agree.

> I was thinking perhaps for simple cases like this one, it is possible to skip ReversePInvokeFrame and optimize it to inlineable return 42.

I think that is minor overhead all things considered. I say that because it is why the example with a blittable signature was so distracting. When factoring in the cost of marshalling non-trivial types, the transition and frames would become far less interesting relatively speaking. If we can see it under a profiler though that would be worth understanding.

The overhead you mention though is related to https://github.com/dotnet/runtime/issues/54107, in the sense of having a way to indicate an `UnmanagedCallersOnly` can skip boilerplate transition with something like `SuppressGCTransition`. This would only make sense in cases where it was a purely blittable API _and_ didn't reference any managed memory. Perhaps there is a better way to express this though or the JIT could compute those cases.

**Comment by jkotas** [comment]
I agree with @AaronRobinsonMSFT that the optimizations of the reverse PInvoke transitions is same problem as #54107. It is something for JIT to do when it is safe. I do not think we would want to make it expressible in public surface.

> This would only make sense in cases where it was a purely blittable API and didn't reference any managed memory.

This also includes calls to any methods, including helpers that the JIT may choose to insert or that may run as side-effect of code execution. 

Also, the code where this optimization kicked in would be impossible to debug with .NET managed debuggers. .NET managed debuggers expect that the thread transitions to cooperative mode before any code runs.

**Comment by jkoritzinsky** [comment]
We aren't going to get to this in .NET 8, moving to .NET 9.

**Comment by ThaDaVos** [comment]
Curious if there are any updates on this since 2023

**Comment by jkoritzinsky** [comment]
I've put some thought into this proposal and here's the design I've come up with:

```csharp
namespace System.Runtime.InteropServices.Marshalling;

[AttributeUsage(AttributeTargets.Method)]
public sealed class LibraryExportAttribute : Attribute
{
    public LibraryExportAttribute();
    public LibraryExportAttribute(bool emitNativeExport); // false -> equivalent behavior to zero-param, true -> same as (string) overload with the name of the method.
    public LibraryExportAttribute(string nativeExportName);

    public StringMarshalling StringMarshalling { get; }
    public Type? StringMarshallingCustomType { get; }
}


[AttributeUsage(AttributeTargets.Method)]
public abstract class UnmanagedEntrypointProviderAttribute : Attribute
{
    static abstract unsafe void* UnmanagedEntrypoint { get; }
}

namespace System.Runtime.InteropServices;

public static class Marshal
{
   public static void* GetUnmanagedEntrypointForMethod(Delegate managedMethod);
}
```

Example usage:

```csharp

public partial class C
{
    [LibraryExport(StringMarshalling = StringMarshalling.Utf8)]
    public static int partial StringLength(string s)
    {
         return s.Length;
    }
}

ReadOnlySpan<byte> str = "Hello World!"u8;
delegate* unmanaged<byte*, int> callback = Marshal.GetUnmanagedEntrypointForMethod(C.StringLength);
fixed (byte* b = str)
{
     callback(str);
}
 
```

The source generator would generate a partial declaration of the attributed method with an attribute of a file-scoped type that derives from `UnmanagedEntrypointProviderAttribute`. The callback would also be in a file-scoped type and would be an `UnmanagedCallersOnly` method that implements the marshalling.

`Marshal.GetUnmanagedEntrypointForMethod` would look up the attribute on the delegate's `Method` property.

To remove the dependency on the reflection stack (and make this usable within CoreLib), we can introduce an interceptor in the generator that directly returns the callback's function pointer for usages that pass a method group and the method is defined in the same assembly (ie the generator is generating the callback and the attribute as well).

**Comment by jkotas** [comment]
Instead of introducing reflection and hoping interceptors will make it go away, would it be better to solve it by convention for the source-generated native entrypoint getter - like a property with `EntryPoint` suffix?
```csharp
public partial class C
{
    [LibraryExport(StringMarshalling = StringMarshalling.Utf8)]
    public static int partial StringLength(string s)
    {
         return s.Length;
    }
}

ReadOnlySpan<byte> str = "Hello World!"u8;
delegate* unmanaged<byte*, int> callback = C.StringLengthEntryPoint;
fixed (byte* b = str)
{
     callback(str);
}
```

**Comment by jkoritzinsky** [comment]
I was following our general guidance of avoiding introducing new user-accessible APIs via generators (ie avoiding introducing possibly public/documentation-requiring surface area).

**Comment by MichalPetryka** [comment]
Another alternative:
```cs
public partial class C
{
    public partial delegate* unmanaged<byte*, int> StringLength { get; }

    [LibraryExport(StringMarshalling = StringMarshalling.Utf8)]
    public static int partial StringLength(string s)
    {
         return s.Length;
    }
}

ReadOnlySpan<byte> str = "Hello World!"u8;
delegate* unmanaged<byte*, int> callback = C.StringLength;
fixed (byte* b = str)
{
     callback(str);
}
```

**Comment by AaronRobinsonMSFT** [comment]
> I was following our general guidance of avoiding introducing new user-accessible APIs via generators (ie avoiding introducing possibly public/documentation-requiring surface area).

I applaud this attempt. It aligns nicely with some tricks created with the COM source generator. However, I am inclined to agree with @jkotas in https://github.com/dotnet/runtime/issues/79786#issuecomment-5040445449. In addition to wanting to avoid reflection and, dear god, interceptors. I think having an opaque API on `Marshal` avoids us being able to validate function pointers. The `Marshal.GetUnmanagedEntrypointForMethod` returns a `void*`, which means users needs to speak the function pointer correctly. If instead it was generated by the source generator or even require users to write it out themselves like in @MichalPetryka's suggestion, we could do validation there as well.

I much prefer the "generate a new member" approach because then we get to define the signature for users and don't need to validate pointless attempts at writing the correct signature. Perhaps we should loop in API design stakeholders and see if the scenario warrants an exception to the rule.

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky We should also think about how we'd handle the native AOT scenario involving [`UnmanagedCallersOnlyAttribute.AssociatedSourceType`](https://github.com/dotnet/runtime/issues/127634).

**Comment by jkoritzinsky** [comment]
I think I'd be ok introducing surface area if the user could specify the name of the method (which should be doable in the attribute). I'll think on a reasonable shape.

> We should also think about how we'd handle the native AOT scenario involving [`UnmanagedCallersOnlyAttribute.AssociatedSourceType`](https://github.com/dotnet/runtime/issues/127634).

For that, I'd recommend that we add an additional property on the attribute that we'd forward down.

**Comment by jkoritzinsky** [comment]
Here's an updated proposal:

```csharp

namespace System.Runtime.InteropServices.Marshalling;

[AttributeUsage(AttributeTargets.Method)]
public sealed class LibraryExportAttribute : Attribute
{
    public LibraryExportAttribute();

    // One or more of the following properties must be set.
	public string? CallbackMethodName { get; set; } // When specified, is the name of the emitted method (otherwise is emitted file-scoped). Must be a valid C# identifier
    public string? UnmanagedExportName { get; set; }

    public Type? AssociatedSourceType { get; set; }

    public StringMarshalling StringMarshalling { get; set; }
    public Type? StringMarshallingCustomType { get; set; }
}

```

There's a few cons with this design:

1. Assigning a function pointer always requires the user to specify the function pointer type (no `var` usage) as `var` would break once an overload is introduced. As a result, the user still needs to specify the whole signature when assigning to a local. Users would be able to Go-to-definition to get the signature so they wouldn't be guessing though, which is an improvement.
2. Unlike the other design, we cannot design a "LibraryExport marshaller" that cleanly gets the callback for a particular delegate instance's underlying method (this marshaller in any case wouldn't work with an interceptor, only the reflection fallback).

**Comment by AaronRobinsonMSFT** [comment]
Another option, similar to @MichalPetryka, but avoids the need for `AssociatedSourceType` and any other `UnmanagedCallersOnly` forwarding. It does require users to write a signature, but it avoids the function pointer syntax at declaration time. It follows the well established `partial` fill-in-the-blank pattern we've been using too, specifically that a method is defined and then generated. We can easily validate the signature and point nicely at a method argument for users to correct as opposed to a function pointer signature that is the special case for a C# signature pattern. 

```csharp
[AttributeUsage(AttributeTargets.Method)]
public sealed class LibraryExportAttribute : Attribute
{
    public LibraryExportAttribute(){}

    // One or more of the following properties must be set.
    public string? UnmanagedExportName { get; set; }

    public StringMarshalling StringMarshalling { get; set; }
    public Type? StringMarshallingCustomType { get; set; }
}

public partial class C
{
    [UnmanagedCallersOnly]
    public static partial int StringLengthUnmanaged(byte* s);

    [LibraryExport(StringMarshalling = StringMarshalling.Utf8, UnmanagedExportName = nameof(StringLengthUnmanaged))]
    public static int StringLength(string s)
    {
         return s.Length;
    }
}
```

**Comment by jkoritzinsky** [comment]
I think that idea makes sense. We could have a code fix that automatically generates the correct signature if the user gets it wrong

**Comment by MichalPetryka** [comment]
> We could have a code fix that automatically generates the correct signature if the user gets it wrong

I think it'd be better to have an analyzer with a codefix for it.

**Comment by am11** [comment]
> `UnmanagedExportName`

Given:

`[LibraryImport("native_lib", EntryPoint = "get_string_length", StringMarshalling = StringMarshalling.Utf8)]`

Since the main motivation was ecosystem consistency, I think `EntryPoint` would fit better:

`[LibraryExport(EntryPoint = nameof(StringLengthUnmanaged), StringMarshalling = StringMarshalling.Utf8)]`

**Comment by MichalPetryka** [comment]
> Since the main motivation was ecosystem consistency, I think `EntryPoint` would fit better:

Keep in mind that `EntryPoint` means something else already for `UnmanagedCallersOnly`.

**Comment by am11** [comment]
Right, I understood that the boilerplate proxy will be taken care by the srcgen.

**Comment by MichalPetryka** [comment]
> Right, I understood that the boilerplate proxy will be taken care by the srcgen.

I've meant that on UCO it means that the method is an unmanaged export from NAOT, so it'd be weird if `EntryPoint` on LibraryExport did not do such an export.

**Comment by am11** [comment]
Not sure how my comment reads like that, what is the point of having LibraryExport which doesn't export anything

**Comment by MichalPetryka** [comment]
> Not sure how my comment reads like that, what is the point of having LibraryExport which doesn't export anything

The idea from @AaronRobinsonMSFT is to have the UCO still be responsible for that and to have the export only do marshalling.

**Comment by kevincathcart-cas** [comment]
So it sounds like all versions of this proposal would include generated reverse P-Invoke-like marshalling, akin to that that `LibraryImport` generates?

It also sounds like any actual exporting would be limited to the scenarios where `UnmanagedCallersOnly` with an entrypoint currently exports?

Those scenarios are currently:
1. NativeAOT
2. Mono AOT "Library Mode"
3. Mono Wasm compilation
4. CoreCLR based Wasm compilation
5. External tooling like DNNE

@AaronRobinsonMSFT 's approach which requires an explicit `[UnmanagedCallersOnly(Entrypoint="...")]` does have the downside of  some asymmetry with `LibraryImport`, but it does have an extra not explicitly mentioned benefit. Specifically it could let the generated marshalling work both for exports, as well as for non-exported native function pointers, depending on if the user specifies an entrypoint on the UCO. That means that approach does not leave behind an export side equivalent to #63590 to figure out later.

**Comment by MichalPetryka** [comment]
> * Mono AOT "Library Mode"
> * Mono Wasm compilation

Those 2 won't apply due to mono being removed in .NET 12
